### ISSUE TITLE
Possible code generation bug compiling ministat(1) with CheriABI

### ISSUE THREAD

**Issue by rwatson** [CTSRD-CHERI/llvm#144]
In attempting to compile ```ministat(1)``` with CheriABI:
```
diff --git a/usr.bin/ministat/Makefile b/usr.bin/ministat/Makefile
index 2a63797..0c7cf0e 100644
--- a/usr.bin/ministat/Makefile
+++ b/usr.bin/ministat/Makefile
@@ -1,6 +1,9 @@
 # $FreeBSD$
 PROG=  ministat
 
+WANT_CHERI=     pure
+WANT_DUMP=      yes
+
 LIBADD=        m
 
 .include <bsd.prog.mk>
```

I see a reproducible problem in which a capability-length exception fires when storing a pointer into a presumed 0x20-size allocation. Digging into the source, we see that ```calloc(3)``` is called to allocate an array of pointers of a suitable size:
```
struct plot {
        double          min;
        double          max;
        double          span;
        int             width;

        double          x0, dx;
        int             height;
        char            *data;
        char            **bar;
        int             separate_bars;
        int             num_datasets;
};

static struct plot plot;

static void
SetupPlot(int width, int separate, int num_datasets)
{
        struct plot *pl;

        pl = &plot;
        pl->width = width;
        pl->height = 0;
        pl->data = NULL;
        pl->bar = NULL;
        pl->separate_bars = separate;
        pl->num_datasets = num_datasets;
        pl->min = 999e99;
        pl->max = -999e99;
}

static void
PlotSet(struct dataset *ds, int val)
{
        struct plot *pl;
        int i, j, m, x;
        unsigned n;
        int bar;

        pl = &plot;
        if (pl->span == 0)
                return;

        if (pl->separate_bars)
                bar = val-1;
        else
                bar = 0;

        if (pl->bar == NULL)
                pl->bar = calloc(sizeof(char *), pl->num_datasets);
        if (pl->bar[bar] == NULL) {
                pl->bar[bar] = malloc(pl->width);
                memset(pl->bar[bar], 0, pl->width);
        }
...
}

...
int
main(int argc, char **argv)
{
...
        if (!flag_n && !suppress_plot) {
                SetupPlot(termwidth, flag_s, nds);
                for (i = 0; i < nds; i++)
                        DimPlot(ds[i]);
                for (i = 0; i < nds; i++)
                        PlotSet(ds[i], i + 1);
                DumpPlot();
        }
...
}
```

Streamtrace confirms that ```SetupPlot``` is run and initialises ```pl->num_datasets``` to 1. In ```PlotSet()```, and that the two functions agree on the location of ```pl``` (which is a pointer to the global variable ```plot``` per above). However, in the code generated to call ```calloc(3)```, it looks like the value of ```pl->num_datasets``` is never loaded: instead, the value 0 is always used, leading to 0 being passed as the size argument to ```malloc(3)```:

```
        if (pl->bar == NULL)
   120004708:   d8750060        clc     c3,zero,96(c21)
   12000470c:   67c1008c        daddiu  at,s8,140
   120004710:   49b85841        csetoffset      c24,c11,at
   120004714:   48950007        cfromptr        c21,c0,zero
   120004718:   49c11d41        cne     at,c3,c21
   12000471c:   1420000c        bnez    at,120004750 <PlotSet+0x168>
   120004720:   67d20090        daddiu  s2,s8,144
                pl->bar = calloc(sizeof(char *), pl->num_datasets);
   120004724:   67818bf8        daddiu  at,gp,-29704
   120004728:   48810047        cfromptr        c1,c0,at
   12000472c:   c8210003        cld     at,zero,0(c1)
   120004730:   48000807        cgetpcc c1
   120004734:   49ac0841        csetoffset      c12,c1,at
   120004738:   64040020        daddiu  a0,zero,32
   12000473c:   48f16000        cjalr   c17,c12
   120004740:   64050000        daddiu  a1,zero,0            <-- Problem here: should be using loaded pl->num_datasets, which is non-zero
```

Earlier initialisation of ```pl->num_datasets``` appears to be as expected:
```
static void
SetupPlot(int width, int separate, int num_datasets)
{
   1200043c8:   67bdffa0        daddiu  sp,sp,-96
   1200043cc:   ebcbeac3        csd     s8,sp,88(c11)
   1200043d0:   03a0f025        move    s8,sp
   1200043d4:   49b96002        cgetoffset      t9,c12
   1200043d8:   3c010008        lui     at,0x8
   1200043dc:   0039082d        daddu   at,at,t9
   1200043e0:   64216228        daddiu  at,at,25128
        struct plot *pl;

        pl = &plot;
        pl->width = width;
   1200043e4:   dc218060        ld      at,-32672(at)
   1200043e8:   64213960        daddiu  at,at,14688
   1200043ec:   48810047        cfromptr        c1,c0,at
   1200043f0:   e88100c2        csw     a0,zero,24(c1)
        pl->height = 0;
   1200043f4:   e8010182        csw     zero,zero,48(c1)
   1200043f8:   48820007        cfromptr        c2,c0,zero
        pl->data = NULL;
   1200043fc:   f8410040        csc     c2,zero,64(c1)
        pl->bar = NULL;
   120004400:   f8410060        csc     c2,zero,96(c1)
        pl->separate_bars = separate;
   120004404:   64010080        daddiu  at,zero,128
   120004408:   e8a10802        csw     a1,at,0(c1)
        pl->num_datasets = num_datasets;
   12000440c:   64010084        daddiu  at,zero,132
   120004410:   e8c10802        csw     a2,at,0(c1)
...
```

Streamtrace seems to confirm that a value of 1 was stored as a result of the call to ```SetupPlot()``` prior to ```PlotSet()```.

**Comment by rwatson** [comment]
Full ```objdump``` of ```ministat(3)```: https://gist.github.com/rwatson/fd7fa99c5c97f34724cc

**Comment by rwatson** [comment]
Various build command lines (preprocessed output to follow).

Depend phase:
```
===> usr.bin/ministat (depend)
CC='/home/rnw24/sdk/sdk//bin/clang -g -integrated-as --target=cheri-unknown-free
bsd  -msoft-float --sysroot=/home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/
tmp -mabi=sandbox -cheri-linker ' mkdep -f ministat.dep_c    -std=gnu99   /home/
rnw24/git/cheribsd/usr.bin/ministat/ministat.c
cat ministat.dep_c > .depend
echo ministat: /home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp/usr/libch
eri/libc.a /home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp/usr/libcheri/
libm.a /home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp/usr/libcheri/libc
.a /home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp/usr/libcheri/libmallo
c_simple.a >> .depend
```

Build phase:
```
===> usr.bin/ministat (all)
/home/rnw24/sdk/sdk//bin/clang -g -integrated-as --target=cheri-unknown-freebsd
 -msoft-float --sysroot=/home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp 
-mabi=sandbox -cheri-linker -O -pipe -G0   -Wno-error  -std=gnu99 -Wsystem-heade
rs -Werror -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -W
missing-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Ws
witch -Wshadow -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wnest
ed-externs -Wredundant-decls -Wold-style-definition -Wno-pointer-sign -Wmissing-
variable-declarations -Wthread-safety -Wno-empty-body -Wno-string-plus-int -Wno-
unused-const-variable -Qunused-arguments -c /home/rnw24/git/cheribsd/usr.bin/min
istat/ministat.c -o ministat.o
/home/rnw24/sdk/sdk//bin/clang -g -integrated-as --target=cheri-unknown-freebsd
 -msoft-float --sysroot=/home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp 
-mabi=sandbox -cheri-linker -O -pipe -G0   -Wno-error  -std=gnu99 -Wsystem-heade
rs -Werror -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -W
missing-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Ws
witch -Wshadow -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wnest
ed-externs -Wredundant-decls -Wold-style-definition -Wno-pointer-sign -Wmissing-
variable-declarations -Wthread-safety -Wno-empty-body -Wno-string-plus-int -Wno-
unused-const-variable -Qunused-arguments  -Wl,--no-warn-mismatch -static -o mini
stat ministat.o  -lm   -lc -lmalloc_simple
Unable to find size of symbol at 012005734c for pointer at 0x1200572c0
Using section size (27347 bytes) instead
Unable to find size of symbol at 0120057352 for pointer at 0x1200572e0
Using section size (27347 bytes) instead
Unable to find size of symbol at 0120057358 for pointer at 0x120057300
Using section size (27347 bytes) instead
Unable to find size of symbol at 012005bf42 for pointer at 0x120057320
Using section size (27347 bytes) instead
Unable to find size of symbol at 01200599d0 for pointer at 0x12005a2c0
Using section size (27347 bytes) instead
Unable to find size of symbol at 01200599dc for pointer at 0x12005a2e0
Using section size (27347 bytes) instead
Unable to find size of symbol at 01200599f4 for pointer at 0x12005a300
Using section size (27347 bytes) instead
Unable to find size of symbol at 0120059a0e for pointer at 0x12005a320
Using section size (27347 bytes) instead
Unable to find size of symbol at 0120059a1e for pointer at 0x12005a340
Using section size (27347 bytes) instead
Unable to find size of symbol at 0120059a36 for pointer at 0x12005a360
Using section size (27347 bytes) instead
Unable to find size of symbol at 0120059a49 for pointer at 0x12005a380
Using section size (27347 bytes) instead
...lots more of these...
Unable to find size of symbol at 012005b739 for pointer at 0x12005b660
Using section size (27347 bytes) instead
Unable to find size of symbol at 012005b7dc for pointer at 0x12005b680
Using section size (27347 bytes) instead
Unable to find size of symbol at 012005b7df for pointer at 0x12005b6a0
Using section size (27347 bytes) instead
strip -o ministat.stripped -w --keep-symbol=__cheri_callee_method.\*  --keep-sym
bol=__cheri_method.\* ministat
gzip -cn /home/rnw24/git/cheribsd/usr.bin/ministat/ministat.1 > ministat.1.gz
objdump -xrsSD ministat > ministat.dump

**Comment by rwatson** [comment]
Preprocessed ```ministat.c``` at: https://gist.github.com/rwatson/4ae2d7ef71f8b86d02f2

Built using the command line:
```
/home/rnw24/sdk/sdk//bin/clang -E -g -integrated-as --target=cheri-unknown-freebsd -msoft-float --sysroot=/home/rnw24/obj/mips.mips64/home/rnw24/git/cheribsd/tmp -mabi=sandbox -cheri-linker -O -pipe -G0 -Wno-error -std=gnu99 -Wsystem-headers -Werror -Wall -Wno-format-y2k -W -Wno-unused-parameter -Wstrict-prototypes -Wmissing-prototypes -Wpointer-arith -Wreturn-type -Wcast-qual -Wwrite-strings -Wswitch -Wshadow -Wunused-parameter -Wcast-align -Wchar-subscripts -Winline -Wnested-externs -Wredundant-decls -Wold-style-definition -Wno-pointer-sign -Wmissing-variable-declarations -Wthread-safety -Wno-empty-body -Wno-string-plus-int -Wno-unused-const-variable -Qunused-arguments -c /home/rnw24/git/cheribsd/usr.bin/ministat/ministat.c -o ministat.preprocessed.c
```

**Comment by rwatson** [comment]
I can confirm that I can reproduce that compiling with ```-O0``` allows the program to execute successfully (or, at least, to completion!):
```
root@beri1:/tmp # ./ministat foo
x foo
+------------------------------------------------------------------------------+
|                                                                             x|
|       x  x  x  x   x                                                        x|
||_______________M______________A_______________________________|              |
+------------------------------------------------------------------------------+
    N           Min           Max        Median           Avg        Stddev
x   7             1            23             4     8.7142857     9.8440216
root@beri1:/tmp # ./ministat foo foo
x foo
+ foo
+------------------------------------------------------------------------------+
|                                                                             *|
|       *  *  *  *   *                                                        *|
||_______________M______________A_______________________________|              |
+------------------------------------------------------------------------------+
    N           Min           Max        Median           Avg        Stddev
x   7             1            23             4     8.7142857     9.8440216
+   7             1            23             4     8.7142857     9.8440216
No difference proven at 95.0% confidence
```

**Comment by davidchisnall** [comment]
It looks as if this was introduced by the fix to #142, which prevented load widening.  Here, the load widening would be safe (loading two 32-bit fields with a 64-bit load).  The fix for this is probably going to be reverting the fix to #142 and fixing it properly.

**Comment by rwatson** [comment]
I can confirm that this appears to be resolved in the cases of ```ncal(1)```, ```ministat(1)```, and ```elfdump(1)```.
